packages/openneuro-server/src/utils/datacite-utils.ts

```diff
@@ -1,61 +1,206 @@
-          `Datacite file for ${datasetId}:${revision} served with unexpected Content-Type: ${contentType}. Attempting YAML parse anyway.`,
-        )
-      }
-
       const text = await res.text()
-      try {
-        // Parse the YAML content into a JavaScript object with the specified type.
-        const parsedYaml: RawDataciteYml = yaml.load(text) as RawDataciteYml
-        return parsedYaml
-      } catch (parseErr) {
-        // If parsing fails, throw a detailed error.
-        throw new Error(
-          `Found datacite file for dataset ${datasetId} (revision: ${revision}), but failed to parse it as YAML:`,
-          { cause: parseErr },
-        )
-      }
+      const parsed: RawDataciteYml = yaml.load(text) as RawDataciteYml
+
+      return parsed
     } else if (res.status === 404) {
-      // If the file is not found (404) return null.
       return null
     } else {
-      // For any other unexpected status code, throw an error.
       throw new Error(
-        `Attempted to read datacite file for dataset ${datasetId} (revision: ${revision}) and received status ${res.status}.`,
+        `Unexpected status ${res.status} when fetching datacite.yml`,
       )
     }
-  } catch (fetchErr) {
-    // Catch any network or other errors during the fetch and report them to Sentry.
-    Sentry.captureException(fetchErr)
+  } catch (err) {
+    Sentry.captureException(err)
     return null
   }
 }
+
+/**
+ * Save datacite.yml back to dataset
+ */
+export const saveDataciteYmlToRepo = async (
+  datasetId: string,
+  userId: string,
+  dataciteData: RawDataciteYml,
+) => {
+  const url = `http://datalad-0/datasets/${datasetId}/files/datacite.yml`
+
+  try {
+    // Directly PUT the file with cookie-based JWT auth
+    await superagent
+      .post(url)
+      .set("Cookie", generateDataladCookie(config)(userId))
+      .set("Accept", "application/json")
+      .set("Content-Type", "text/yaml")
+      .send(yaml.dump(dataciteData))
+
+    // Commit the draft after upload
+    const gitRef = await commitFiles(datasetId, userId)
+    return { id: gitRef }
+  } catch (err) {
+    Sentry.captureException(err)
+    throw err
+  }
+}
+/**
+ * Converts RawDataciteContributor -> internal Contributor type.
+ * Optionally attaches a `userId` if the contributor exists as a site user.
+ */
+export const normalizeRawContributors = async (
+  raw: RawDataciteContributor[] | undefined,
+): Promise<Contributor[]> => {
+  if (!Array.isArray(raw)) return []
+
+  const orcids = raw
+    .map((c) => validateOrcid(c.nameIdentifiers?.[0]?.nameIdentifier))
+    .filter(Boolean) as string[]
+
+  const users = await User.find({ orcid: { $in: orcids } }).exec()
+  const orcidToUserId = new Map(users.map((u) => [u.orcid, u.id]))
+
+  return raw.map((c, index) => {
+    const contributorOrcid = validateOrcid(
+      c.nameIdentifiers?.[0]?.nameIdentifier,
+    )
+    return {
+      name: c.name ||
+        [c.givenName, c.familyName].filter(Boolean).join(" ") ||
```

Comment:
Might be good to follow the suggested order from the datacite spec which is "Family Name, Given Name".